python/ffsim/hamiltonians/diagonal_coulomb_hamiltonian.py

```diff
@@ -0,0 +1,171 @@
+                }
+            )
+        for p, q in itertools.product(range(self.norb), repeat=2):
+            (coeff_same_spin, coeff_diff_spin) = 0.5 * self.diag_coulomb_mats[:, p, q]
+            op += FermionOperator(
+                {
+                    (cre_a(p), des_a(p), cre_a(q), des_a(q)): coeff_same_spin,
+                    (cre_a(p), des_a(p), cre_b(q), des_b(q)): coeff_diff_spin,
+                    (cre_b(p), des_b(p), cre_a(q), des_a(q)): coeff_diff_spin,
+                    (cre_b(p), des_b(p), cre_b(q), des_b(q)): coeff_same_spin,
+                }
+            )
+        return op
+
+    @staticmethod
+    def from_fermion_operator(op: FermionOperator) -> DiagonalCoulombHamiltonian:
+        """Convert a FermionOperator to a DiagonalCoulombHamiltonian."""
+
+        # extract norb
+        norb = 1 + max(orb for term in op for _, _, orb in term)
+
+        # initialize variables
+        constant: float = 0
+        one_body_tensor = np.zeros((norb, norb), dtype=complex)
+        diag_coulomb_mats = np.zeros((2, norb, norb), dtype=float)
+
+        # populate the tensors
+        for term, coeff in op.items():
+            if not term:  # constant term
+                constant = coeff.real
+            elif len(term) == 2:  # one-body term candidate
+                (_, _, p), (_, _, q) = term
+                terms = [(cre_a(p), des_a(q)), (cre_b(p), des_b(q))]
+                if term in terms:
+                    one_body_tensor[p, q] += 0.5 * coeff
+                else:
+                    raise ValueError(
+                        "FermionOperator cannot be converted to "
+                        "DiagonalCoulombHamiltonian"
+                    )
+            elif len(term) == 4:  # two-body term candidate
+                (_, _, p), (_, _, _), (_, _, q), (_, _, _) = term
+                terms_same_spin = [
+                    (cre_a(p), des_a(p), cre_a(q), des_a(q)),
+                    (cre_b(p), des_b(p), cre_b(q), des_b(q)),
+                ]
+                terms_diff_spin = [
+                    (cre_a(p), des_a(p), cre_b(q), des_b(q)),
+                    (cre_b(p), des_b(p), cre_a(q), des_a(q)),
+                ]
+                if term in terms_same_spin:
+                    diag_coulomb_mats[0][p, q] += coeff.real
+                elif term in terms_diff_spin:
+                    diag_coulomb_mats[1][p, q] += coeff.real
+                else:
+                    raise ValueError(
+                        "FermionOperator cannot be converted to "
+                        "DiagonalCoulombHamiltonian"
+                    )
+            else:
+                raise ValueError(
+                    "FermionOperator cannot be converted to DiagonalCoulombHamiltonian"
+                )
+
+        # ensure diag_coulomb_mats symmetry
+        for i in range(2):
+            diag_coulomb_mats[i] = 0.5 * (diag_coulomb_mats[i] + diag_coulomb_mats[i].T)
```

Comment:
Use the `transpose` function from Numpy and do it in place to avoid allocating more memory than needed.
```suggestion
        diag_coulomb_mats += diag_coulomb_mats.transpose(0, 2, 1)
        diag_coulomb_mats *= 0.5
```

Reply:
Nice!